src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+            ;; record how many uSTX this pox-addr will stack for in the given reward cycle
+            (append-reward-cycle-pox-addr
+                (get pox-addr args)
+                reward-cycle
+                (get amount-ustx args))
+
+            ;; update running total
+            (map-set reward-cycle-total-stacked
+                { reward-cycle: reward-cycle }
+                { total-ustx: (+ (get amount-ustx args) total-ustx) }
+            )
+            
+            ;; updated _this_ reward cycle
+            u1
+        ))
+        u0
+    ))
+)
+
+;; Add a PoX address to a given sequence of reward cycle lists.
+;; A PoX address can be added to at most 12 consecutive cycles.
+;; No checking is done.
+;; Returns the number of reward cycles added
+(define-private (add-pox-addr-to-reward-cycles (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+                                               (first-reward-cycle uint)
+                                               (num-cycles uint)
+                                               (amount-ustx uint))
+    (begin
+        ;; For safety, add up the number of times (add-principal-to-ith-reward-cycle) returns 1.
+        ;; It _should_ be equal to num-cycles.
+        (asserts! (is-eq num-cycles (fold +
+           (map add-pox-addr-to-ith-reward-cycle (list
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u0 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u1 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u2 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u3 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u4 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u5 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u6 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u7 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u8 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u9 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u10 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u11 }))
+           u0))
```

Comment:
Should should be able to make this block more concise, with something like:
```
(define-private (do-something (cycle-index uint) (params (tuple (arg1 int) (arg2 int) (num-cycles uint))))
    (begin
        ;; so something
        { arg1: (get arg1 params), arg2: (get arg2 params), num-cycles: (+ (get num-cycles params) u1) }))

(let ((cycle-indexes (list u1 u2 u3 u4 u5 u6 u7 u8 u9 u10 u11 u12)))
      (fold do-something cycle-indexes { arg1: 10, arg2: 10, num-cycles: u0 }))
```

Reply:
I think this is hard to do concisely without anonymous functions -- we'd need to be able to "pass in" the `pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx` fields to the applied function.

Reply:
Isn't what's happening in my snippet? it's passing-in the args by using a tuple as an accumulator, and only mutating the component `num-cycle` when the operation is successful.

Reply:
I agree - that would be easier with anonymous function

Reply:
Ah, yes, I see now

Reply:
> I think this is hard to do concisely without anonymous functions

Partial function application for `map`, `fold` and `filter` is one option, like in [Pact](https://pact-language.readthedocs.io/en/stable/pact-reference.html#functional-concepts), providing some of the power of anonymous functions without going all the way to _closures._

Another possibility is building on the variadic `map` proposed in #1482, with the repeated parameters passed in as separate sequence(s). A restricted `repeat` function could be used to make all sequence arguments to `map` have the same length, if not done implicitly for conciseness.

Reply:
@lgalabru Sure, I can restructure this so the inner function returns the next inner function invocation's arguments.  Thanks for the tip! :+1:

Reply:
@njordhov Thank you for your comments, but with all due respect, this is not the time or place for proposing changes to the Clarity language.  I have to work with what Clarity offers _right now_ ;)

Reply:
>  this is not the time or place for proposing changes to the Clarity language. I have to work with what Clarity offers right now ;)

That's the nature of flying the plane while you build it. While this PR obviously are limited to what Clarity currently supports, such use cases are a good opportunity to identify potential improvements to the language. I'll add a note to #1482 for reference.

Reply:
@jcnelson Proposal posted at https://github.com/clarity-lang/reference/issues/27. Thank you for the illustrative use case.

Reply:
Related proposals: https://github.com/blockstack/stacks-blockchain/issues/1844

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/7a8aed196cbf329dbd4c92ec004b1db7b713f411